deploy/olm-catalog/hawtio-operator/0.3.0/hawtio-operator.v0.3.0.clusterserviceversion.yaml

```diff
@@ -0,0 +1,275 @@
+    mediatype: image/svg+xml
+  install:
+    spec:
+      clusterPermissions:
+      - rules:
+        - apiGroups:
+          - oauth.openshift.io
+          resources:
+          - oauthclients
+          verbs:
+          - '*'
+        - apiGroups:
+          - config.openshift.io
+          resources:
+          - clusterversions
+          verbs:
+          - get
+        - apiGroups:
+          - apiextensions.k8s.io
+          resources:
+          - customresourcedefinitions
+          verbs:
+          - get
+          - list
+          - watch
+        - apiGroups:
+          - console.openshift.io
+          resources:
+          - consolelinks
+          verbs:
+          - create
+          - delete
+          - get
+          - list
+          - patch
+          - update
+          - watch
+        serviceAccountName: hawtio-operator
+      deployments:
+      - name: hawtio-operator
+        spec:
+          replicas: 1
+          selector:
+            matchLabels:
+              name: hawtio-operator
+          strategy: {}
+          template:
+            metadata:
+              labels:
+                name: hawtio-operator
+            spec:
+              containers:
+              - command:
+                - hawtio-operator
+                env:
+                - name: WATCH_NAMESPACE
+                  valueFrom:
+                    fieldRef:
+                      fieldPath: metadata.namespace
+                - name: POD_NAME
+                  valueFrom:
+                    fieldRef:
+                      fieldPath: metadata.name
+                - name: OPERATOR_NAME
+                  value: hawtio-operator
+                - name: IMAGE_REPOSITORY
+                  value: registry.redhat.io/fuse7/fuse-console@sha256:04d9eea074ce3985cd8da6d7613c2ec8f544b205ba1f43fa59f035e9ef373cf6
```

Comment:
The `IMAGE_REPOSITORY` environment variable is designed to hold the image without the tag/digest. The later comes from the hawtio CR `.spec.version` field, as it's possible for users to change the Hawtio / Fuse console version the operator deploys.

One option may be to set the retrieved digest into the CSV example and console sample. Or to have a variable that can be set with the retrieved digest instead of using the `latest` tag by default.

Still I wonder what it means to be able to change the version in an offline environment. WDYT?

Reply:
> I'd suggest we do not use the product image here upstream, but rather use them in the product repository https://github.com/jboss-fuse/fuse-console-operator, by regenerating the CSV with product image there.

https://github.com/jboss-fuse/fuse-console-operator - link is broken and does this mean I need to open the PR there? 




> The `IMAGE_REPOSITORY` environment variable is designed to hold the image without the tag/digest. The later comes from the hawtio CR `.spec.version` field, as it's possible for users to change the Hawtio / Fuse console version the operator deploys.
> 
> One option may be to set the retrieved digest into the CSV example and console sample. Or to have a variable that can be set with the retrieved digest instead of using the `latest` tag by default.
> 
> Still I wonder what it means to be able to change the version in an offline environment. WDYT?

the other around I am thinking to introduce a new boolean variable like UseImageTags/UseImageDigest in CR and add extra environment variable IMAGE_DIGEST in csv file- base on CR input it will use tag or digest.

Reply:
> https://github.com/jboss-fuse/fuse-console-operator - link is broken and does this mean I need to open the PR there?

The repository is private. I'll have you added ASAP. It rebuilds the upstream operator with different options and contains the product CSV. So I think that's OK to work here for now, and once it gets merged, we can think of updating the product repo.

> the other around I am thinking to introduce a new boolean variable like UseImageTags/UseImageDigest in CR and add extra environment variable IMAGE_DIGEST in csv file- base on CR input it will use tag or digest.

My understanding is that the `version` field from the CR can be used to specify either a tag or a digest. The problem is that currently, if the `version` field is not present, it defaults to `latest`. It seems we could introduce a Go build variable, like we do for `ImageRepository`: https://github.com/hawtio/hawtio-operator/blob/4183906c78c01df6e981de33682c3493cfcb51c1/pkg/controller/hawtio/hawtio_controller.go#L73 I think it'd be logical as the digest is associated to the image repository. Then we would not need to introduce the `HAWTIO-IMAGE` environment variable. The only difficulty I see is that the tooling that retrieves the digest would have to be compiled before compiling the operator code. WDYT?

Reply:
> > https://github.com/jboss-fuse/fuse-console-operator - link is broken and does this mean I need to open the PR there?
> 
> The repository is private. I'll have you added ASAP. It rebuilds the upstream operator with different options and contains the product CSV. So I think that's OK to work here for now, and once it gets merged, we can think of updating the product repo.
> 
> > the other around I am thinking to introduce a new boolean variable like UseImageTags/UseImageDigest in CR and add extra environment variable IMAGE_DIGEST in csv file- base on CR input it will use tag or digest.
> 
> My understanding is that the `version` field from the CR can be used to specify either a tag or a digest. The problem is that currently, if the `version` field is not present, it defaults to `latest`. It seems we could introduce a Go build variable, like we do for `ImageRepository`:
> 
> https://github.com/hawtio/hawtio-operator/blob/4183906c78c01df6e981de33682c3493cfcb51c1/pkg/controller/hawtio/hawtio_controller.go#L73
> 
> I think it'd be logical as the digest is associated to the image repository. Then we would not need to introduce the `HAWTIO-IMAGE` environment variable. The only difficulty I see is that the tooling that retrieves the digest would have to be compiled before compiling the operator code. WDYT?

how the operator will identify that version contains digest or tag because inside code is appending - repository + ":" + tag, this good for using tag but for digest we required to add @.

 and what in case if the user enters/change incorrect version/digest in CR

Reply:
> how the operator will identify that version contains digest or tag because inside code is appending - repository + ":" + tag, this good for using tag but for digest we required to add @.

I guess we can detect the version is a digest, assuming it follows `hash_algo:hash`. Or we can imagine mandating the user to provide the separator, e.g., `:latest` or `@sha256:...`.

> and what in case if the user enters/change incorrect version/digest in CR

This is already a point today. The `version` field is used to support upgrade. I wonder as well what upgrade means in offline mode?